src/ucs/datastruct/mpool.c

```diff
@@ -265,6 +283,19 @@ void ucs_mpool_grow(ucs_mpool_t *mp, unsigned num_elems)
 void *ucs_mpool_get_grow(ucs_mpool_t *mp)
 {
     ucs_mpool_data_t *data = mp->data;
+    size_t max_chunk_size, chunk_size;
+    unsigned num_elems;
+
+    if (data->chunks != NULL) {
+        /* add 0.5 for roundup */
+        num_elems      = data->elems_per_chunk * data->grow_factor + 0.5;
+        max_chunk_size = data->max_chunk_size;
+        chunk_size     = ucs_mpool_chunk_size(mp, num_elems);
+        if (chunk_size <= max_chunk_size) {
+            /* growing and check quota */
+            data->elems_per_chunk = ucs_min(data->quota, num_elems);
```

Comment:
To save some memory for cases such as left quota e.g. =1 when the chunk will grow to 2048(num of elements).
But the original behaviour will allocate specified num_elems. If this is not worthy of the change, I can remove it.

Reply:
For "logic of grow_factor should be moved into ucs_mpool_grow() function.", I did not get well.
For API: void ucs_mpool_grow(ucs_mpool_t *mp, unsigned num_elems); There are other 2 places calling it besides ucs_mpool_get_grow().
1) https://github.com/openucx/ucx/blob/e01c3eeb01a1282a5abcaa90dd60c37dd00043e4/src/uct/ib/ud/base/ud_iface.c#L567
2) https://github.com/openucx/ucx/blob/e01c3eeb01a1282a5abcaa90dd60c37dd00043e4/test/gtest/ucs/test_mpool.cc#L202
It seemed the caller wanted to alloc the specified num_elems.
If move the logic of grow_factor into this ucs_mpool_grow(), should remove the 2nd parameter(num_elems)?
If keep the API not changed, which num_elems(the parameter or mp->data->elems_per_chunk) should be used?

Reply:
For "need to apply the max_chunk_size limit in ucs_mpool_grow, not here.
so the chunk size can grow really up to max_chunk_size, and not less than that (because of the factor)."
I did not get it. Could you explain for which case(factor=?) it would grow less than max_chunk_size and it is not as expected?

Reply:
> To save some memory for cases such as left quota e.g. =1 when the chunk will grow to 2048(num of elements).

This logic should be in `ucs_mpool_grow` because no matter where it's called from, it still has to enforce the quota

> For "need to apply the max_chunk_size limit in ucs_mpool_grow, not here.

Something like this:

```c
ucs_mpool_grow()
{
    ...
    chunk_size = ucs_min(chunk_size, data->max_chunk_size);
    ...
}
```


```c
ucs_mpool_get_grow()
{
    ...
    ucs_mpool_grow(mp, data->elems_per_chunk);
    if (mp->freelist == NULL) {
        return NULL;
    }

    ucs_assert(data->chunks != NULL);

    /* Stop growing if last chunk reached the limit */
    num_elems             = ucs_min(data->elems_per_chunk,
                                    data->chunks->num_elems);
    data->elems_per_chunk = (num_elems * data->grow_factor) + 0.5;

    return ucs_mpool_get(mp);
}
```

Reply:
Thank you, now I'm clear with the idea. Just two minor changes based on above code:
1) the smaller one should always be "data->chunks->num_elems" which is the real num of this latest chunk, right?
     num_elems             = ucs_min(data->elems_per_chunk,
                                    data->chunks->num_elems);
2) is for this comment, this is calculating num of elements for next growing, right?
    /* Stop growing if last chunk reached the limit */

Reply:
1. NO because it's possible the number of allocated elements would be larger because of alignment (when not capped by max_chunk_size)
2. Yes, since data->elems_per_chunk is the number of elements for next chunk allocation

Reply:
Got it and fixed it.